app/models/miq_server/worker_management/monitor/kubernetes.rb

```diff
@@ -0,0 +1,100 @@
+module MiqServer::WorkerManagement::Monitor::Kubernetes
+  extend ActiveSupport::Concern
+  attr_accessor :pod_resource_version
+
+  def current_pods
+    @current_pods ||= Concurrent::Hash.new
+  end
+
+  def reset_current_pods
+    @current_pods = nil
+    current_pods
+  end
+
+  def cleanup_failed_deployments
+    ensure_pod_monitor_started
+
+    # TODO: We should have a list of worker deployments we'll delete to avoid accidentally killing pg/memcached/orchestrator
+    # See ContainerOrchestrator#get_pods
+    failed_deployments.each do |failed|
+      orchestrator.delete_deployment(failed)
+    end
+  end
+
+  def failed_deployments(restart_count = 5)
+    # TODO: This logic might flag deployments that are hitting memory/cpu limits or otherwise not really 'failed'
+    current_pods.select { |_name, h| h.fetch(:last_state_terminated) && h.fetch(:container_restarts, 0) > restart_count }.collect { |_name, h| h[:label_name] }
+  end
+
+  private
+  def start_pod_monitor
+    @monitor_thread ||= begin
+      reset_current_pods
+      Thread.new { monitor_pods }
+    end
+  end
+
+  def ensure_pod_monitor_started
+    if @monitor_thread.nil? || !@monitor_thread.alive?
+      if !@monitor_thread.nil? && @monitor_thread.status.nil?
+        dead_thread, @monitor_thread = @monitor_thread, nil
+        _log.info("#{log_prefix} Waiting for the Monitor Thread to exit...")
+        dead_thread.join
+      end
+
+      start_pod_monitor
+    end
```

Comment:
copied from similar behavior in the [event catcher](https://github.com/ManageIQ/manageiq/blob/681ab0ab268cd603cc19c56cb425b6b934194366/app/models/manageiq/providers/base_manager/event_catcher/runner.rb#L191-L201)

Reply:
As discussed with you and @agrare, maybe not for this PR, but I suggest we move the general pattern of having a WatchThread that can be auto-restarted into a generic for in the core manageiq repo in the lib dir, for eventual extraction into kubeclient.  This way issues that arise can be fixed in one place (like the 410 gone issue).

I recommened it lives in core, and then the kubernetes / openshift providers use it directly.  We may need to tweak the interface to be less ems oriented and more generic.